sonar-db/src/main/java/org/sonar/db/CollationChecker.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.db;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.CheckForNull;
+import org.apache.commons.lang.StringUtils;
+import org.picocontainer.Startable;
+import org.sonar.api.utils.MessageException;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.db.dialect.MsSql;
+import org.sonar.db.dialect.MySql;
+import org.sonar.db.dialect.Oracle;
+import org.sonar.db.dialect.PostgreSql;
+
+import static java.lang.String.format;
+import static org.apache.commons.lang.StringUtils.containsIgnoreCase;
+import static org.apache.commons.lang.StringUtils.endsWithIgnoreCase;
+
+/**
+ * SONAR-6171
+ * Check that database has UTF8 character set and case-sensitive collation.
+ * As obviously tables must be checked after being created, this component
+ * must not be executed at the same time as {@link DatabaseChecker}.
+ */
+public class CollationChecker implements Startable {
+
+  private final Database db;
+  private final StatementExecutor statementExecutor;
+
+  public CollationChecker(Database db) {
+    this(db, new StatementExecutor());
+  }
+
+  @VisibleForTesting
+  CollationChecker(Database db, StatementExecutor statementExecutor) {
+    this.db = db;
+    this.statementExecutor = statementExecutor;
+  }
+
+  @Override
+  public void start() {
+    try {
+      Loggers.get(getClass()).info("Verify database collation");
+      check();
+    } catch (SQLException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Override
+  public void stop() {
+    // nothing to do
+  }
+
+  private void check() throws SQLException {
+    try (Connection connection = db.getDataSource().getConnection()) {
+      switch (db.getDialect().getId()) {
```

Comment:
you should add a default which throws a `IllegalArgumentException` to handle at runtime when a new value ID is added but no implementation of check has been provided

Of course, this is impossible to test cover :)